### DISCUSSION TITLE
Option to use Voltage instead of charge?

### DISCUSSION THREAD

**Discussion by derekoharrow** [andrew-codechimp/HA-Battery-Notes#681] [Ideas]
### Checklist

- [X] I have filled out the template to the best of my ability.
- [X] This only contains 1 feature request (if you have multiple feature requests, open one feature request for each feature request).
- [X] This issue is not a duplicate feature request of [previous feature requests](https://github.com/andrew-codechimp/ha-battery-notes/issues?q=is%3Aopen+is%3Aissue+label%3Aenhancement).

### Is your feature request related to a problem? Please describe.

Some devices are bad at reporting the state of charge - especially things like Aqara PIRs, etc.

Many devices will, however, report their voltage, which can be used to infer the battery state - e.g. Voltage dropping from 3,005mV to, say, 2,300mV

### Describe the solution you'd like

It would be great to have an option to use the Voltage instead of Charge state for specific devices.

Perhaps, a device specific configuration item that says to use voltage instead of charge?

### Describe alternatives you've considered

n/a

### Additional context

An example of a problematic device:

I have a Xiaomi GZCGQ01LM Lux Sensor. This is reporting the battery state as 0%, and hence battery low. However, the Voltage is being reported as 2,700mV and the device is still working perfectly, and has been for many months.

This would be a great example of where Voltage is a battery indication of battery health than charge.

**Comment by andrew-codechimp** [comment]
Thanks for the explanation and example, I'll have a think about whether I can handle this in the future.
I am loathe to introduce a lot of config options per device but I'll see what I can come up with.

**Comment by andrew-codechimp** [comment]
Also related
Some devices use a binary sensor of device_class battery to indicate battery health - e.g. Tado TRVs

HA Battery Notes doesn't seem to work with these.

Describe the solution you'd like

For devices using a battery device class binary_sensor, translate the on/off (Normal/Low) battery state to the Battery Low state for Battery Notes. The Battery Notes battery low binary sensor should match the device battery sensor.

Describe alternatives you've considered

None

Additional context

With Tado TRV devices as an example, they have a binary_sensor for battery state. HA Battery Notes doesn't seem to provide the battery low binary sensor for these devices. HA says "The entity is no longer being provided by the battery_notes integration..."

**Comment by andrew-codechimp** [comment]
Thinking about these non-standard batteries I'll be continually playing catch up with all the variants.
A thought I have is on the device configuration you can optionally specify a template on how to get the battery, that way anyone could write a jinja template to create the percentage.
Will have to think about how the value is updated though.

**Reply by derekoharrow** [reply]
Hopefully there aren't too many variants - device_class of battery is one which perhaps should be covered as standard is it's a built-in device class for HA.

Template would be great for other variants like measuring voltage instead of charge. I guess the template would need to return an on/off or true/false value to say if the battery is ok or not, which would be quite simple for measuring voltage, e.g.:

{{ states('sensor.lux_sensor_voltage')|int < 2500 }}

**Reply by andrew-codechimp** [reply]
That would be the plan, it would also cope with the battery low status some devices have.
A little bit of effort to create for each sensor but a lot of flexibility.
That's if I can work out how to evaluate a template of course, it's on the list of things to investigate soon.

**Comment by mihsu81** [comment]
@andrew-codechimp Would it be possible to have the Battery Low binary sensors of the type Battery, without having to modify it manually?
It seems to have been created for this particular case.
Not really clear from the discussions I've read so far.
Thanks.

**Reply by andrew-codechimp** [reply]
Can you give me an example of a device with battery_low, I've not come across these myself. 
Is it an entity or attribute?

**Reply by mihsu81** [reply]
It looks like this with the default _device_class_, which i think is _None_:
![image](https://github.com/andrew-codechimp/HA-Battery-Notes/assets/40715497/38b7cfaa-8ea5-4896-99ef-df12a6a7dcbc)

And like this with the _device_class_ _Battery_ set manually:
![image](https://github.com/andrew-codechimp/HA-Battery-Notes/assets/40715497/59fe858f-4a69-4141-9d02-538eb448c380)
![image](https://github.com/andrew-codechimp/HA-Battery-Notes/assets/40715497/cb1616a4-3177-4d57-a85c-3e89c6058234)

**Reply by andrew-codechimp** [reply]
Ah, I see what you mean now. 
I don't think that would be a problem but I'll do some tests.

**Comment by allthesebugsv2** [comment]
I'm not sure whether this also fits in here: I have some devices which report battery status as an enum.

For example this guy here: https://www.zigbee2mqtt.io/devices/ZTH08-E.html

In will report low, medium, high for battery state and is currently not picked up by Battery Notes.
![Screenshot 2024-02-01 134523](https://github.com/andrew-codechimp/HA-Battery-Notes/assets/50304079/bdb6a051-fdf6-47c9-adcb-3bcb66f77193)

**Reply by andrew-codechimp** [reply]
That would work with what I have planned, you'd just create a template that returns true if enum is low.

**Comment by andrew-codechimp** [comment]
This feature is coming along so I thought I'd share an early view of what I'll be providing.

Here's the draft doc
https://github.com/andrew-codechimp/HA-Battery-Notes/blob/low-templates/docs/index.md#battery-low-template

And a screenshot of the config
![CleanShot 2024-02-13 at 15 47 57@2x](https://github.com/andrew-codechimp/HA-Battery-Notes/assets/1849731/99ad096a-268a-4ade-aefa-11b296098fef)

**Reply by derekoharrow** [reply]
Simple, but effective - that would do nicely!!

**Reply by andrew-codechimp** [reply]
Thanks for the feedback. Hoping to release next week bundled with another small feature once I've done some more testing.

**Reply by blalor** [reply]
I'm happy to see this. The docs already reflect the change, but the version installed from HACS doesn't support it, yet; that was confusing but I guess it'll sort itself out, soon! 😆

**Reply by andrew-codechimp** [reply]
This is in the beta release (2.1.x) this morning so if you turn on beta's within HACS you'll get it.
I haven't really found a good way to manage updating the docs for betas only that won't add effort.

**Reply by derekoharrow** [reply]
Just playing with it now and have my first device setup with a template.
Battery Low is now reporting correctly based on a device voltage. I notice that Battery % is still at 0, which I presume is normal and that we should now use Battery Low as an indicator of low battery rather than %age?

**Reply by blalor** [reply]
I wonder if it'd play better with the existing threshold and alerts/events if the template returned a percentage or level instead of a binary?

**Reply by andrew-codechimp** [reply]
The threshold event still fires with the low template, I guess the voltage could be converted to a percentage but it would make the other use cases harder to template.

**Comment by andrew-codechimp** [comment]
> Just playing with it now and have my first device setup with a template. Battery Low is now reporting correctly based on a device voltage. I notice that Battery % is still at 0, which I presume is normal and that we should now use Battery Low as an indicator of low battery rather than %age?

Did Battery+ show a percentage before this release?

**Reply by derekoharrow** [reply]
Not sure TBH.

I might have picked a bad example though as the battery on that device is reporting as 0% anyway (despite it still working).
Other similar devices are showing a battery % and the battery+ sensor is showing the same, so I guess all is well.

But, to @blalor 's point, if using the template to override the high/low sensor, what does one do with the %age sensor...

**Reply by andrew-codechimp** [reply]
Good on the battery+ sensor being OK, my esp device was emulating all the scenarios and I didn't see that dropping the battery+.

I think the intention would be to create a template that returned something like
80% when voltage is high, 10% when voltage is low
80% when enum = good, 20% when enum is low
Then Battery+ could show those figures and you could use the global/device threshold values to trigger the low.
I like that it would use the thresholds but for any device with a battery %, but you prefer to use the voltage, it would be confusing.
